### PR TITLE
Bug 1311150 - Prevent multiple highlights appearing from the same domain

### PR DESCRIPTION
**Pull request by bkmunar**

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
ButtonToast.swift | 18 | Colons should be next to the identifier when specifying a type. 
#### Errors

File | Line | Reason |
| --- | ----- | ----- |
ButtonToast.swift | 23 | Opening braces should be preceded by a single space and on the same line as the declaration. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: Storage/SQL/SQLiteHistoryRecommendations.swift [hunk 1]
```diff
@@ -50,6 +50,10 @@ extension SQLiteHistory: HistoryRecommendations {
             "       ORDER BY visitDate DESC" +
             "   )" +
             "   LEFT JOIN \(TableHistory) ON \(TableHistory).id = s" +
+            "   INNER JOIN (" +
+            "       SELECT \(TableHistory).domain_id AS domain_id, MAX(\(TableHistory).id) AS history_id FROM \(TableHistory)" +
+            "       GROUP BY \(TableHistory).domain_id" +
+            "   ) AS domains ON domains.domain_id = \(TableHistory).domain_id AND \(TableHistory).id = history_id" +
```
Comment (sleroux) [inline Storage/SQL/SQLiteHistoryRecommendations.swift]: Using `MAX` here for the query to return a unique, predictable result makes sense but I think we might be able to improve this by using the last visit date instead since ID is somewhat arbitrary. Turns out we already have a view named `ViewHistoryVisits` that we can query that will tell us this. The only thing missing from the view is the domain ID but I believe we can surface that in the view from the history table we're joining with. When adding the column to the view we'll need to make sure in the migration that we DROP the old view and recreate it. We do this a few times throughout our versions so you can see how we do it there.

On a minor note, since we're using the same derived table subquery in both the `nonRecentHistory` and `bookmarkHighlights`, we can DRY this up by moving the query string into a variable and inserting it in the query.

Comment (sleroux) [inline Storage/SQL/SQLiteHistoryRecommendations.swift]: @rnewman might have some other thoughts - I'm not the most SQL-savvy 😦

File: StorageTests/TestSQLiteHistoryRecommendations.swift [hunk 2]
```diff
@@ -189,6 +189,61 @@ class TestSQLiteHistoryRecommendations: XCTestCase {
...
+        history.addLocalVisit(siteVisitC3).succeeded()
+
+        history.addLocalVisit(siteVisitD1).succeeded()
+        history.addLocalVisit(siteVisitD2).succeeded()
+        history.addLocalVisit(siteVisitD3).succeeded()
+        history.addLocalVisit(siteVisitD4).succeeded()
+
+        let highlights = history.getHighlights().value.successValue!
+        print(highlights.count)
+        XCTAssertEqual(highlights.count, 1)
```
Comment (sleroux) [inline StorageTests/TestSQLiteHistoryRecommendations.swift]: If we amend the derived table subquery to using `MAX` on the visit date, we should add a test case to verify that the returned unique result is always the latest.

Comment (sleroux) [inline StorageTests/TestSQLiteHistoryRecommendations.swift]: Nit: Also, since this test case just tests removing duplicate highlights, I would clean this up a bit and don't bother adding non-facebook.com sites just to make the case more clear.

File: Storage/SQL/SQLiteHistoryRecommendations.swift (hunk 1 above)
Comment (rnewman) [inline Storage/SQL/SQLiteHistoryRecommendations.swift]: If you use the date, you'd need to join again to get the contributing grouped value, or sub select with an order and limit.

**Comment by rnewman** [comment]
As noted on IRC, a solution like this would indeed ensure that we only return one highlight for each domain… but it doesn't address the question of *which result we should return* very thoroughly.

That is: if there are multiple candidates for a given domain, which one should we choose?

(And there's a broader question: if I have two highlights from Medium.com, or two from Wikipedia, perhaps we should show both instead of some lower-scoring highlight from another domain…)

I can think of various approaches:

- Choose the most recently visited. (That's the patch as written, I think.)
- Choose the least recently visited within a timeframe — resurfacing old hits.
- Choose the one never visited on this device — space-shifting.
- Choose the least visited item, count-wise — something you missed.
- Use some other kind of score: e.g., the item for which we have the best imagery, or it's in our cache, or it has a known title — optimize for display.
- … and so on.

In short: this is a product question.

**Comment by bkmunar** [comment]
@sleroux back to you! We're going with the most recently visited for now as per Barbara's request -- will file another bug to change if there is a more definite product decision

**Comment by sleroux** [comment]
LGTM 👍 My highlights are already looking much better!
